app/src/lib/git/tag.ts

```diff
@@ -62,16 +63,28 @@ export async function fetchTagsToPush(
     '--porcelain',
   ]
 
-  const result = await git(args, repository.path, 'fetchTagsToPush', {
-    env: await envForRemoteOperation(account, remote.url),
-  })
+  let result: IGitResult | null = null
 
-  const lines = result.stdout.split('\n')
+  try {
+    result = await git(args, repository.path, 'fetchTagsToPush', {
+      env: await envForRemoteOperation(account, remote.url),
+      successExitCodes: new Set([0, 1]),
+    })
+  } catch (e) {
+    // Even if we receive an unexpected error code, we try to
+    // do the best effort to try to parse the stdout to find unpushed tags.
+    // If we can't even parse the stdout, we return an empty array of unpushed tags.
+    if (e instanceof GitError) {
+      result = e.result
+    }
+  }
+
+  const lines = result ? result.stdout.split('\n') : []
```

Comment:
I'm not sure returning an empty list for error cases is a good idea with the cacheing in this PR. It seems like we should not cache an error result, or at least not cache error results that are empty lists.

@rafeca do we have a sense of what kind of errors we might encounter here, or is this a general precaution?

Reply:
Really good point! ✨  This is meant to silently handle at least two kind of expected errors (but there may be more):

1. Current branch is not fully in sync with the remote one (3rd point in @tierninho's [comment](https://github.com/desktop/desktop/pull/9576#issuecomment-616902677)).
2. Computer not connected to Internet.

I see the issue of caching an error result, which may get resolved over time while the cached result will be outdated with the empty list. Still, if we don't cache anything when we have these errors, it's possible that we end making too many dry-run pushes in these scenarios...

Good news is that the current cache handles scenario 1. (the cache key contains the HEAD commit sha, so once the branch gets synced the cache will get invalidated), but scenario 2 or other potential errors will result in outdated results. I see a few options here:

1. Don't cache any error result (this will cause a lot of operation when having persistent errors).
2. Keep caching error results (this will cause outdated results when having temporary errors).
3. Use a smarter cache (like [`node-lru-cache`](https://github.com/isaacs/node-lru-cache) which allows us to set a max age for entries. This won't completely solve any of the problems of the other two approaches but can be in a nice sweet spot to mitigate both quite reasonably. 

Given that 3. will also solve https://github.com/desktop/desktop/pull/9576/files#r416242898 I'm going to opt to implement that one, if you feel strongly about it we can reconsider it.

Reply:
this looks great! i haven't used `node-lru-cache` before and i'm excited to.